common/FastJmp.cpp

```diff
@@ -0,0 +1,86 @@
+/*  PCSX2 - PS2 Emulator for PCs
+ *  Copyright (C) 2002-2021  PCSX2 Dev Team
+ *
+ *  PCSX2 is free software: you can redistribute it and/or modify it under the terms
+ *  of the GNU Lesser General Public License as published by the Free Software Found-
+ *  ation, either version 3 of the License, or (at your option) any later version.
+ *
+ *  PCSX2 is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+ *  PURPOSE.  See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with PCSX2.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "FastJmp.h"
+
+#ifndef _WIN32
+
+#if defined(__x86_64__)
+
+asm(
+	".global fastjmp_set\n\t"
+	".global fastjmp_jmp\n\t"
+	".text\n\t"
+	"\n\t"
+	"fastjmp_set:\n\t"
+	"movq 0(%rsp), %rax\n\t"
+	"movq %rsp, %rdx\n\t" // fixup stack pointer, so it doesn't include the call to fastjmp_set
+	"addq $8, %rdx\n\t"
+	"movq %rax, 0(%rdi)\n\t" // actually rip
+	"movq %rbx, 8(%rdi)\n\t"
+	"movq %rdx, 16(%rdi)\n\t" // actually rsp
+	"movq %rbp, 24(%rdi)\n\t"
+	"movq %r12, 32(%rdi)\n\t"
+	"movq %r13, 40(%rdi)\n\t"
+	"movq %r14, 48(%rdi)\n\t"
+	"movq %r15, 56(%rdi)\n\t"
+	"xorl %eax, %eax\n\t"
+	"ret\n\t"
+	"\n\t"
+	"fastjmp_jmp:\n\t"
+	"movl %esi, %eax\n\t"
+	"movq 0(%rdi), %rdx\n\t" // actually rip
+	"movq 8(%rdi), %rbx\n\t"
+	"movq 16(%rdi), %rsp\n\t" // actually rsp
+	"movq 24(%rdi), %rbp\n\t"
+	"movq 32(%rdi), %r12\n\t"
+	"movq 40(%rdi), %r13\n\t"
+	"movq 48(%rdi), %r14\n\t"
+	"movq 56(%rdi), %r15\n\t"
+	"jmp *%rdx\n\t");
```

Comment:
Since we have fancy C++ things, thoughts on using multiline strings for this?  That way you don't need all the `\n\t"` https://gcc.godbolt.org/z/eY448c53c

Assembler doesn't seem to take `//` though, so you'l have to use `#` or `/* */`

Reply:
Didn't end up being *as* clean because I had to add the function prefix in there for macs, but it's a bit easier on the eyes now at least.